crates/electrum/src/electrum_ext.rs

```diff
@@ -305,7 +305,7 @@ fn construct_update_tip(
     // Atomically fetch the latest `ASSUME_FINAL_DEPTH` count of blocks from Electrum. We use this
     // to construct our checkpoint update.
     let mut new_blocks = {
-        let start_height = new_tip_height.saturating_sub(ASSUME_FINAL_DEPTH);
+        let start_height = (new_tip_height + 1).saturating_sub(ASSUME_FINAL_DEPTH);
```

Comment:
```suggestion
        let start_height = new_tip_height.saturating_sub(ASSUME_FINAL_DEPTH - 1);
```

I think this is more correct.

Reply:
Yes, assume final depth can be 0.

Reply:
If `ASSUME_FINAL_DEPTH` is 0, we still want to grab the tip (otherwise there will be no chain update). If `ASSUME_FINAL_DEPTH` is 1, we assume the tip can be reorged so I'll be convenient to grab the prev block also.

```rust
let start_height = new_tip_height.saturating_sub(ASSUME_FINAL_DEPTH);
let hashes = client
    .block_headers(start_height as _, (ASSUME_FINAL_DEPTH + 1) as _)?
```

I think this logic is better.

Reply:
> If `ASSUME_FINAL_DEPTH` is 0, we still want to grab the tip (otherwise there will be no chain update).

If this is `CHAIN_SUFFIX_LENGTH`, conceptually if it was 0 you should just not fetch the tip and use the checkpoint passed in as the tip (i.e. 0 means we should not extend the chain). But this is useless and not much point reasoning about it since it's an internal constant we control. 

So in my view your code is wrong because it returns `ASSUME_FINAL_DEPTH + 1` blocks but we only want to return `ASSUME_FINAL_DEPTH` blocks (again this should be changed to `CHAIN_SUFFIX_LENGTH`). Or was this intentional?

Reply:
> Also can we change the name of `ASSUME_FINAL_DEPTH` to just `CHAIN_SUFFIX_LENGTH`. I don't think this code is "assuming" anything is final. It's just including a chain suffix of a certain length for the purpose of robustness as far as I understand.

But we are assuming something. The reason why we fetch multiple blocks in an atomic/batched call is to ensure consistency between the fetched blocks (as there is a still a slim chance that a reorg happens when we are fetching blocks individually). After the batch call, we can find the "point of agreement" by fetching lower blocks individually, since there is no chance of reorg as according to our `ASSUME_FINAL_DEPTH`.

---

**I think we should have the following:**

* `ASSUME_FINAL_DEPTH` should be called `MAX_REORG_DEPTH`. Where `MAX_REORG_DEPTH` of 0 means the tip is final.
* The reason why we want to fetch `MAX_REORG_DEPTH + 1` number of blocks in the first batch call, is we want to at least include one block which is "final". This is beneficial if we anchor the tx to the lowest-possible block.

Reply:
I implemented the above mentioned here: 5a902d5e0fefa9da5cd5635496dbbed1314884f3

Reply:
> But we are assuming something. The reason why we fetch multiple blocks in an atomic/batched call is to ensure consistency between the fetched blocks (as there is a still a slim chance that a reorg happens when we are fetching blocks individually). After the batch call, we can find the "point of agreement" by fetching lower blocks individually, since there is no chance of reorg as according to our `ASSUME_FINAL_DEPTH`.

I claim this algorithm does not assume what you claim. The slim chance that a reorg happens while you are fetching blocks does not matter since you check the tip again later on. If the tip has changed then you throw away everything and start again. If we are truly assuming something explain what changes when the assumption breaks down i.e. there is a reorg greater than `MAX_REORG_DEPTH`.  What happens differently as opposed to reorgs under that depth?

The assumption that the algorithm makes is that if block X existed one time and then it still exists at a later time then it existed the whole time in between. This is not a great assumption but it's what we can do with the API we have but under this assumption the algorithm is robust under a re-org of any depth.

Reply:
@LLFourn fair point.